### PR TITLE
parallel stages

### PR DESCRIPTION
**Pull request by robfletcher**
This implements the graph we need to make parallel stages work.

@ajordens PTAL I'm going to work on getting actual stages using this now but see if you think this is a reasonable model of how they'll look.

### CODE REVIEW

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/BranchingStageDefinitionBuilder.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Collection;
+import java.util.Map;
+import com.netflix.spinnaker.orca.pipeline.model.Execution;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+
+/**
+ * Implement for stages that will create parallel branches to perform the same
+ * tasks for multiple contexts. For example, a multi-region bake or deploy.
+ */
+public interface BranchingStageDefinitionBuilder extends StageDefinitionBuilder {
```
Comment (robfletcher) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/BranchingStageDefinitionBuilder.java]: Stages that need parallel stuff will implement this

File: orca-bakery/src/test/groovy/com/netflix/spinnaker/orca/bakery/pipeline/BakeStageSpec.groovy [hunk 2]
```diff
@@ -121,7 +121,8 @@ class BakeStageSpec extends Specification {
     }
   }
 
-  private static List<Map> expectedContexts(String amiSuffix, String... regions) {
+  private
+  static List<Map> expectedContexts(String amiSuffix, String... regions) {
```
Comment (ajordens) [inline orca-bakery/src/test/groovy/com/netflix/spinnaker/orca/bakery/pipeline/BakeStageSpec.groovy]: accidental line break

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java [hunk 3]
```diff
@@ -37,99 +45,154 @@ public ExecutionRunnerSupport(Collection<StageDefinitionBuilder> stageDefinition
...
-   * @param stage the stage with no tasks currently attached.
-   * @param <T>   the execution type.
+   * @param stage    the stage with no tasks currently attached.
+   * @param callback callback invoked with the tasks of the stage.
+   * @param <T>      the execution type.
    */
-  protected <T extends Execution<T>> void planStage(Stage<T> stage) {
+  protected <T extends Execution<T>> void planStage(
+    Stage<T> stage,
+    // TODO: surely we can just use the 2nd consumer form here and handle single-element
```
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java]: Certainly

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java [hunk 4]
```diff
@@ -37,99 +45,154 @@ public ExecutionRunnerSupport(Collection<StageDefinitionBuilder> stageDefinition
...
+    }
+    task.setId(String.valueOf((stage.getTasks().size() + 1)));
+    task.setName(taskDef.getName());
+    task.setStatus(NOT_STARTED);
+    task.setImplementingClass(taskDef.getImplementingClass());
+    if (isEnd) {
+      if (isSubGraph) {
+        task.setLoopEnd(true);
+      } else {
+        // TODO: may need to revisit this for parallel stages like bake & deploy
```
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java]: Is this still relevant?

(I'm not sure that it is ... I suspect we still should be setting `stageEnd` even for the internally parallel stages like bake & deploy)

**Comment by ajordens** [comment]
I think it's a reasonable API ... LGTM.

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java (hunk 4 above)
Comment (robfletcher) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java]: I tried it out and yeah, we still need it. The issue is closed but I think we need a dependency rev to get the fix.

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java (hunk 3 above)
Comment (robfletcher) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java]: Yeah, this is kind of the one shitty thing I want to clean up before merging this.

**Pull request merged**